eth/db/journal.py

```diff
@@ -24,237 +26,230 @@ class DeletedEntry:
-        return changeset_id in self.journal_data
+    def has_changeset(self, changeset_id: JournalDBCheckpoint) -> bool:
+        return changeset_id in self._journal_data
 
-    def record_changeset(self, custom_changeset_id: uuid.UUID = None) -> uuid.UUID:
+    def record_changeset(
+            self,
+            custom_changeset_id: JournalDBCheckpoint = None) -> JournalDBCheckpoint:
         """
-        Creates a new changeset. Changesets are referenced by a random uuid4
+        Creates a new changeset. Changesets are referenced by a random int
         to prevent collisions between multiple changesets.
         """
         if custom_changeset_id is not None:
-            if custom_changeset_id in self.journal_data:
+            if custom_changeset_id in self._journal_data:
                 raise ValidationError(
                     "Tried to record with an existing changeset id: %r" % custom_changeset_id
                 )
             else:
                 changeset_id = custom_changeset_id
         else:
-            changeset_id = uuid.uuid4()
+            changeset_id = next(id_generator)
 
-        self.journal_data[changeset_id] = {}
+        self._journal_data[changeset_id] = {}
         return changeset_id
 
-    def pop_changeset(self, changeset_id: uuid.UUID) -> Dict[bytes, Union[bytes, DeletedEntry]]:
-        """
-        Returns all changes from the given changeset.  This includes all of
-        the changes from any subsequent changeset, giving precedence to
-        later changesets.
-        """
-        if changeset_id not in self.journal_data:
-            raise KeyError(changeset_id, "Unknown changeset in JournalDB")
-
-        all_ids = tuple(self.journal_data.keys())
-        changeset_idx = all_ids.index(changeset_id)
-        changesets_to_pop = all_ids[changeset_idx:]
-        popped_clears = tuple(idx for idx in changesets_to_pop if idx in self._clears_at)
-        if popped_clears:
-            last_clear_idx = changesets_to_pop.index(popped_clears[-1])
-            changesets_to_drop = changesets_to_pop[:last_clear_idx]
-            changesets_to_merge = changesets_to_pop[last_clear_idx:]
-        else:
-            changesets_to_drop = ()
-            changesets_to_merge = changesets_to_pop
+    def discard(self, through_checkpoint_id: JournalDBCheckpoint) -> None:
+        for _ in range(len(self._journal_data)):
+            checkpoint_id, rollback_data = self._journal_data.popitem()
+
+            for old_key, old_value in rollback_data.items():
+                if old_value is REVERT_TO_WRAPPED:
+                    # The current value may not exist, if it was a delete followed by a clear,
+                    # so pop it off, or ignore if it is already missing
+                    self._current_values.pop(old_key, None)
+                elif old_value is DELETE_WRAPPED:
+                    self._current_values[old_key] = old_value
+                else:
```

Comment:
implicit else?  I recognize this is performance code so no big deal.  Might still be worth doing `elif type(old_value) is bytes` which should be decently quick (maybe as-fast as the `cast` call which might be able to be done away with if `mypy` is smart enough.

Reply:
Discard is used rarely enough that I'm good adding this validation back in. The `diff()` is way too sensitive, so I'll leave it out there.